In [0]:
import numpy as np
import pandas as pd

### Q. 연속된 정수 1 ~ 24 와 25~50  으로 각각 구성된 (8,3) 행렬 데이터프레임 df_a와 df_b를 생성 출력하세요.

In [8]:
df_a = pd.DataFrame(np.array(np.arange(1, 25)).reshape(8,3))
df_b = pd.DataFrame(np.array(np.arange(26, 50)).reshape(8,3))

print( df_a, '\n')
print( df_b)

    0   1   2
0   1   2   3
1   4   5   6
2   7   8   9
3  10  11  12
4  13  14  15
5  16  17  18
6  19  20  21
7  22  23  24 

    0   1   2
0  26  27  28
1  29  30  31
2  32  33  34
3  35  36  37
4  38  39  40
5  41  42  43
6  44  45  46
7  47  48  49


### Q. df_a, df_b의 사칙연산 결과를 데이터프레임으로 합쳐서 출력하세요.

In [11]:
add = df_a + df_b
sub = df_a - df_b
mul = df_a * df_b
div = round(df_a / df_b, 2)

print (add)
print (sub)
print (mul)
print (div)

    0   1   2
0  27  29  31
1  33  35  37
2  39  41  43
3  45  47  49
4  51  53  55
5  57  59  61
6  63  65  67
7  69  71  73
    0   1   2
0 -25 -25 -25
1 -25 -25 -25
2 -25 -25 -25
3 -25 -25 -25
4 -25 -25 -25
5 -25 -25 -25
6 -25 -25 -25
7 -25 -25 -25
      0     1     2
0    26    54    84
1   116   150   186
2   224   264   306
3   350   396   444
4   494   546   600
5   656   714   774
6   836   900   966
7  1034  1104  1176
      0     1     2
0  0.04  0.07  0.11
1  0.14  0.17  0.19
2  0.22  0.24  0.26
3  0.29  0.31  0.32
4  0.34  0.36  0.38
5  0.39  0.40  0.42
6  0.43  0.44  0.46
7  0.47  0.48  0.49


### Q. df_a,df_b 데이터 프레임으로 아래 사항을 수행하여 df_na와 df_nb를 생성하세요.
- 4행에서 끝행까지 선택 추출한 후 df_an, df_bn으로 각각 저장
- df_an, df_bn의 인덱스를 각각 리셋한 후 df_na, df_nb로 저장
- df_na의 세번째 행 x 위치에 np.nan 적용
- df_nb의 네번째 행, z 위치에 np.nan 적용
- df_na, df_nb를 출력, NaN 값 확인

In [40]:
df_an = df_a[4:] ; df_bn = df_b[4:]
print(df_an) ; print(df_bn) ; print ()

df_na = df_an.reset_index(drop=True)
df_nb = df_bn.reset_index(drop=True)
print (df_na)
print (df_nb)

df_na.iloc[2] = np.nan
df_nb.iloc[3] = np.nan

print (df_na)
print (df_nb)

    0   1   2
4  13  14  15
5  16  17  18
6  19  20  21
7  22  23  24
    0   1   2
4  38  39  40
5  41  42  43
6  44  45  46
7  47  48  49

    0   1   2
0  13  14  15
1  16  17  18
2  19  20  21
3  22  23  24
    0   1   2
0  38  39  40
1  41  42  43
2  44  45  46
3  47  48  49
      0     1     2
0  13.0  14.0  15.0
1  16.0  17.0  18.0
2   NaN   NaN   NaN
3  22.0  23.0  24.0
      0     1     2
0  38.0  39.0  40.0
1  41.0  42.0  43.0
2  44.0  45.0  46.0
3   NaN   NaN   NaN


### Q. df_na와 df_nb의 NaN 값을 0으로 대체하고 사칙연산을 수행하세요.

In [41]:
add = df_na.add(df_nb, fill_value=0)
sub = df_na.sub(df_nb, fill_value=0)
mul = df_na.mul(df_nb, fill_value=0)
div = df_na.div(df_nb, fill_value=0)

print (add)
print (sub)
print (mul)
print (div)

      0     1     2
0  51.0  53.0  55.0
1  57.0  59.0  61.0
2  44.0  45.0  46.0
3  22.0  23.0  24.0
      0     1     2
0 -25.0 -25.0 -25.0
1 -25.0 -25.0 -25.0
2 -44.0 -45.0 -46.0
3  22.0  23.0  24.0
       0      1      2
0  494.0  546.0  600.0
1  656.0  714.0  774.0
2    0.0    0.0    0.0
3    0.0    0.0    0.0
          0         1         2
0  0.342105  0.358974  0.375000
1  0.390244  0.404762  0.418605
2  0.000000  0.000000  0.000000
3       inf       inf       inf


### Q. 타이타닉 데이터인 train.csv 파일을 데이터프레임으로 변환 후 아래 작업을 수행하세요.

* 다운받은 파일을 jupyter notebook 작업폴더에 저장한다.
* pd.read_csv를 이용 읽기를 수행하고 titanic_df 이름으로 저장한 후 처음 5행을 출력한다.

In [47]:
titanic_df = pd.read_csv('train.csv')
print (titanic_df.head())
print (type(titanic_df))
print (titanic_df.keys())

   PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]
<class 'pandas.core.frame.DataFrame'>
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


### Q. titanic_df에서 Age, Fare 칼럼 선택해서 titanic_af로 출력

In [51]:
titanic_af = titanic_df[['Age', 'Fare']]
print (titanic_af)

      Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]


### Q. 데이터프레임 titanic_af에 숫자 10 더한 후 'addition' 이름으로 데이터프레임을 생성하고 출력하세요.
### Q. 데이터프레임 addition에서 숫자 titanic_af 뺀 후 'subtraction' 이름으로 데이터프레임을 생성하고 출력하세요.

In [53]:
addition = titanic_af + 10
print (addition)
subtraction = addition-titanic_af
print (subtraction)

      Age     Fare
0    32.0  17.2500
1    48.0  81.2833
2    36.0  17.9250
3    45.0  63.1000
4    45.0  18.0500
..    ...      ...
886  37.0  23.0000
887  29.0  40.0000
888   NaN  33.4500
889  36.0  40.0000
890  42.0  17.7500

[891 rows x 2 columns]
      Age  Fare
0    10.0  10.0
1    10.0  10.0
2    10.0  10.0
3    10.0  10.0
4    10.0  10.0
..    ...   ...
886  10.0  10.0
887  10.0  10.0
888   NaN  10.0
889  10.0  10.0
890  10.0  10.0

[891 rows x 2 columns]


### Q. titanic_df의 인덱스를 PassengerId로 바꾸세요.

In [55]:
titanic_df_idx = titanic_df.set_index('PassengerId')
print (titanic_df_idx.head())

             Survived  Pclass  ... Cabin Embarked
PassengerId                    ...               
1                   0       3  ...   NaN        S
2                   1       1  ...   C85        C
3                   1       3  ...   NaN        S
4                   1       1  ...  C123        S
5                   0       3  ...   NaN        S

[5 rows x 11 columns]


### Q. titanic_df의 행 인덱스를 정수형 위치 인덱스로 초기화 하세요.

In [56]:
titanic_df_idx.reset_index(inplace=True)
print(titanic_df_idx)

     PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]


### Q. titanic_df를 Name 칼럼을 기준으로 올림차순 정렬한 후 titanic_sort 이름으로 저장후 출력하세요.

In [62]:
titanic_sort = titanic_df.sort_values('Name')
print (titanic_sort.head()) ; print ()
print (titanic_sort['Name'])

     PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
845          846         0       3  ...   7.55   NaN         S
746          747         0       3  ...  20.25   NaN         S
279          280         1       3  ...  20.25   NaN         S
308          309         0       2  ...  24.00   NaN         C
874          875         1       2  ...  24.00   NaN         C

[5 rows x 12 columns]

845                      Abbing, Mr. Anthony
746              Abbott, Mr. Rossmore Edward
279         Abbott, Mrs. Stanton (Rosa Hunt)
308                      Abelson, Mr. Samuel
874    Abelson, Mrs. Samuel (Hannah Wizosky)
                       ...                  
286                  de Mulder, Mr. Theodore
282                de Pelsmaeker, Mr. Alfons
361                del Carlo, Mr. Sebastiano
153          van Billiard, Mr. Austin Blyler
868              van Melkebeke, Mr. Philemon
Name: Name, Length: 891, dtype: object


### Q. titanic_df에서 Sex' 칼럼 분리후 value_counts를 출력하세요.

In [68]:
titanic_sex = titanic_df['Sex']
print (titanic_sex.value_counts())

male      577
female    314
Name: Sex, dtype: int64


### Q. titanic_sex를 DataFrame으로 변환하여 출력하세요.

In [71]:
titanic_sex = pd.DataFrame(titanic_sex)
print (type(titanic_sex), '\n', titanic_sex)

<class 'pandas.core.frame.DataFrame'> 
         Sex
0      male
1    female
2    female
3    female
4      male
..      ...
886    male
887  female
888  female
889    male
890    male

[891 rows x 1 columns]


### Q. titanic_df에서 'Survived' 칼럼을 추출해서 titanic_survived 이름으로 저장후 자료형을 확인하세요.
### Q. titanic_survived의 처음 5개행의 값을 출력하세요.(범위 ':'을 이용)

In [74]:
titanic_survived = titanic_df['Survived']
print (type(titanic_survived), '\n', titanic_survived[:6])

<class 'pandas.core.series.Series'> 
 0    0
1    1
2    1
3    1
4    0
5    0
Name: Survived, dtype: int64


### Q. 데이터프레임 sex_df와 시리즈 titanic_survived를 추가한 후 ss_df 이름으로 저장 후 출력하세요. 

### Q. ss_df에 890행을 복사해서 891 행에 추가한 후 끝에서 부터 5개행을 출력하세요. 

### Q. ss_df에서 행 인덱스 891과 열 인덱스 첫번째에 해당하는 값을 loc, iloc를 활용해서 각각 구한 후 출력하세요.

### Q. titanic_df의 칼럼을 확인한 후 'PassengerId'와 'Survived' 를 제외한 모든 칼럼을 포함한 df_c2를 출력하세요.

### (응용)Q. train.csv 데이터 셋을 탐색한 후 타이타닉 예측 모델 개발을 위한 분석용 데이터 셋을 만들기 위하여 필요한 전처리 작업을 기술하세요.